engine/src/main/java/com/google/android/fhir/sync/Sync.kt

```diff
@@ -264,30 +249,45 @@ object Sync {
 
   private fun createSyncState(
     workInfoState: WorkInfo.State,
-    currentSyncJobStatus: SyncJobStatus,
-    terminalSyncJobStatus: SyncJobStatus,
-  ): SyncState {
-    return when (workInfoState) {
-      ENQUEUED -> {
-        Enqueued
-      }
-      RUNNING -> {
-        Running(currentSyncJobStatus)
-      }
-      SUCCEEDED -> {
-        Succeeded(terminalSyncJobStatus)
-      }
-      FAILED -> {
-        Failed(terminalSyncJobStatus)
-      }
-      CANCELLED -> {
-        Cancelled
-      }
-      else -> {
-        Unknown
+    syncJobStatusFromWorkManager: SyncJobStatus?,
+    syncJobStatusFromDataStore: SyncJobStatus?,
+  ): SyncState =
+    when (syncJobStatusFromWorkManager) {
+      is SyncJobStatus.Started,
+      is SyncJobStatus.InProgress, -> Running(syncJobStatusFromWorkManager)
+      null -> handleNullWorkManagerStatus(workInfoState, syncJobStatusFromDataStore)
+      else -> error("Error message here")
```

Comment:
i think it might need to be like this? what do you think?
```suggestion
      is SyncJobStatus.Started,
      is SyncJobStatus.InProgress, -> Running(syncJobStatusFromWorkManager)
      is SyncJobStatus.Finished -> Succeeded(syncJobStatusFromWorkManager
      else -> handleNullWorkManagerStatus(workInfoState, syncJobStatusFromDataStore)
```

Reply:
I do not see any advantages if we update the existing code with the suggested change. However, at the same time, I also do not see any issues if we accommodate the suggested change. The reason is that the terminal state (finished, failed) is only fetched from the dataStore, so it will not be part of in-progress states. If the build has legacy code and gets upgraded to the current pull request, in that case, the terminal syncJobStatus gets updated through setProgress, but this state never gets published, and this is the original issue where the terminal state gets lost.

Let me test this behavior.